### ISSUE TITLE
Invalid type error when accessing sanitized `Optional[Callable]` in lambda

### ISSUE THREAD

**Issue by plafosse** [microsoft/pyright#2016]
The following produces a type error improperly. 
```python
from typing import Callable, Optional

def p(func:Optional[Callable[[], None]]=None):
    if func is not None:
        x = lambda: func()
```

Giving the following error
```
Object of type "None" cannot be called
```

**VS Code extension or command-line**
Using VS Code extension using version 1.1.149

**Comment by JelleZijlstra** [comment]
This is strictly speaking correct, because you could do this:
```python
from typing import Callable, Optional

def p(func:Optional[Callable[[], None]]=None):
    if func is not None:
        x = lambda: func()
    func = None  # allowed by the type!
    x()  # fails at runtime
```
However, this code seems unlikely in practice, so it may be more pragmatic to apply type narrowing in nested functions too.

**Comment by erictraut** [comment]
A lambda represents a different execution scope from the outer function. A type checker cannot generally know when a lambda is going to be executed. (It could be handed to another function and queued for later execution.) For that reason, types are not narrowed for symbols captured by this inner scope. Jelle's example demonstrates why a type checker cannot make this assumption.

The recommended workaround is to assign the validated value to a local variable.

```python
def p(func: Optional[Callable[[], None]] = None):
    if func is not None:
        valid_func = func
        x = lambda: valid_func()
```

**Comment by plafosse** [comment]
Ah got it, thank you for the explanation.

**Comment by erictraut** [comment]
Reopening this issue and marking it fixed. After thinking about it more, I found a way to detect when it's safe to apply type narrowing for captured local variables and parameters. See https://github.com/microsoft/pyright/issues/2343 for details.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.172, which I just published. It will also be in the next release of pylance.
